compiler/rustc_middle/src/ty/subst.rs

```diff
@@ -208,6 +233,17 @@ pub type InternalSubsts<'tcx> = List<GenericArg<'tcx>>;
 pub type SubstsRef<'tcx> = &'tcx InternalSubsts<'tcx>;
 
 impl<'a, 'tcx> InternalSubsts<'tcx> {
+    /// Checks whether all elements of this list are types, if so, transmute.
+    pub fn try_as_type_list(&'tcx self) -> Option<&'tcx List<Ty<'tcx>>> {
+        if self.iter().all(|arg| matches!(arg.unpack(), GenericArgKind::Type(_))) {
```

Comment:
Maybe checking the pointer tag directly would make it possible for LLVM to optimize this? Something like `self.iter().all(|arg| arg.ptr.get() & TAG_MASK == TYPE_TAG)`? 

Adding `#[inline]` to this method is a good idea in any case.